File: leetcode/algorithms/tree/binary-tree-preorder-traversal-iterative.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function preorderTraversal(root: TreeNode | null): number[] {
    
    const stack: TreeNode[] = [], trav: number[] = [];
    
    if(root!==null){
     stack.push(root);   
    }
    
    while(stack.length!==0){
        const currNode = stack.pop();
        
        trav.push(currNode.val);
        
        if(currNode.right!==null){
            stack.push(currNode.right);
        }
        if(currNode.left!==null){
            stack.push(currNode.left)
        }
    }
    
    return trav;

};

File: leetcode/algorithms/tree/binary-tree-preorder-traversal-iterative.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        
        if(root==null){
            return new ArrayList();
        }
        
        Stack<TreeNode> elemStack = new Stack<TreeNode>();
        List<Integer> intList = new ArrayList<Integer>();
        
        elemStack.push(root);
        
        while(!elemStack.isEmpty()){
            TreeNode nextElem = elemStack.pop();
            if(nextElem.right!=null)elemStack.push(nextElem.right);
            if(nextElem.left!=null)elemStack.push(nextElem.left);
            intList.add(nextElem.val);
        }
        
        return intList;
    }
}